### ISSUE TITLE
Feature Request: Make ZFS again Ceph compatible (max rados name size)

### ISSUE THREAD

**Issue by RubenKelevra** [zfsonlinux/zfs#4913]
Currently there is a limit of 256 Byte per Filename, like discussed here: https://github.com/zfsonlinux/zfs/issues/2344

This is an issue with some usecases like ceph.

Please add this as a feature-request to increase the maximum filename to get sufficient headroom for every usecase.

A typical error message by ceph while starting on zfs:

```
server131 systemd[1]: Starting Ceph object storage daemon...
server131 systemd[1]: ceph-osd@0.service holdoff time over, scheduling restart.
server131 systemd[1]: ceph-osd@0.service failed.
server131 systemd[1]: Unit ceph-osd@0.service entered failed state.
server131 systemd[1]: ceph-osd@0.service: main process exited, code=exited, status=1/FAILURE
server131 ceph-osd[4055]: 2016-08-01 11:33:03.097621 7f6d1d292800 -1  ** ERROR: osd init failed: (36) File name too long
server131 ceph-osd[4055]: 2016-08-01 11:33:03.082899 7f6d1d292800 -1 osd.0 0 (36) File name too long
server131 ceph-osd[4055]: 2016-08-01 11:33:03.082894 7f6d1d292800 -1 osd.0 0    osd max object namespace len = 256
server131 ceph-osd[4055]: 2016-08-01 11:33:03.082892 7f6d1d292800 -1 osd.0 0    osd max object name len = 2048
server131 ceph-osd[4055]: 2016-08-01 11:33:03.082882 7f6d1d292800 -1 osd.0 0 backend (filestore) is unable to support max object name[space] len
server131 ceph-osd[4055]: 2016-08-01 11:33:02.895242 7f6d1d292800 -1 filestore(/var/lib/ceph/osd/ceph-0) WARNING: max attr value size (1024) is smaller than osd_max_object_name_len (2048).  Your backend filesystem appears to not support attrs large enough to handle the configured max rados name size.  You may
get unexpected ENAMETOOLONG errors on rados operations or buggy behavior
server131 ceph-osd[4055]: starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /dev/disk/by-partlabel/journal-0
server131 systemd[1]: Started Ceph object storage daemon.
server131 ceph-osd-prestart.sh[4005]: create-or-move updated item name 'osd.0' weight 5.2075 at location {host=server131,root=default} to crush map
```

**Comment by tuxoko** [comment]
It doesn't seems like it's talking about filename. Can you do strace to see what exactly it's doing?

**Comment by RubenKelevra** [comment]
@tuxoko actually it is talking about this issue:

"**Your backend filesystem appears to not support attrs large enough to handle the configured max rados name size.**"

This is documented here:
http://docs.ceph.com/docs/jewel/rados/configuration/filesystem-recommendations/#not-recommended

Actually this could be also read as xattr-naming is too long, I'm actually unsure, but running ceph in the latest version on zfs is currently a bit tricky and not fully supported, as this document states:

"This may result in difficult-to-diagnose errors if you try to use RGW or other librados clients that do not properly handle or politely surface any resulting ENAMETOOLONG errors."

Actually I didn't want to go back to filesystems like XFS because of this issue :(

**Comment by RubenKelevra** [comment]
I've renamed it to reduce the confusion. :)

**Comment by behlendorf** [comment]
Yes, please add the output of a strace trace since we shouldn't be hitting any limits on the ZFS side.

**Comment by RubenKelevra** [comment]
@behlendorf whats about this? https://github.com/zfsonlinux/zfs/issues/2344#issuecomment-47282688

**Comment by behlendorf** [comment]
The warning suggests the issue is the maximum xattr size and not the name length.  ZFS supports xattrs up to 64K in size and that limit comes from the Linux kernel and not ZFS, internally ZFS has no limit.  We really need to see the exact `setxattr()` command issued by ceph.

```WARNING: max attr value size (1024) is smaller than osd_max_object_name_len (2048).```

**Comment by RubenKelevra** [comment]
@kolbrich kannst du auf einer Testumgebung das mal ausführen?

---

We'll try it out and send you a report back asap. This might take a while for us.

**Comment by lorddoskias** [comment]
Chiming in, since I have some experience with Ceph. Bear in mind that Ceph is VERY xattr heavy, to the point that ext4 as an underlying system is a no-no: https://bugzilla.kernel.org/show_bug.cgi?id=107301

**Comment by kevin-olbrich** [comment]
Offtopic:

Ruben and I use ZFS for the permanent part of Ceph, the log is written to an SSD capable of handling the load. We are using Ceph for RBDs on OpenStack, currently we don't experience xattr-problems but this could be the case for CephFS. Also the bug report dates back to 2015 where Ceph Hammer release was published which was replaced by infernalis and jewel (both performance optimization releases).

I did not test releases before hammer but with our current infrastructure we did not experience problems - just this warnung.

**Comment by behlendorf** [comment]
Also off topic:

For anyone using the latest ZFS master code and Ceph I'd encourage you to try out the recently merged large dnode feature 50c957f702ea6d08a634e42f73e8a49931dd8055.  It should improve performance for xattr heavy workloads like Ceph.  You will probably want to experiment a little with what the best dnodesize is for Ceph.  I'd suggest trying to set it large enough to store all the xattrs for the common case   if possible.

**Comment by dweeezil** [comment]
I grabbed the ceph sources and it's pretty clear what they're doing.  Its idea of the maximum size of an xattr is hardcoded per-filesystem as follows:
```
// max xattr value size
OPTION(filestore_max_xattr_value_size, OPT_U32, 0)      //Override
OPTION(filestore_max_xattr_value_size_xfs, OPT_U32, 64<<10)
OPTION(filestore_max_xattr_value_size_btrfs, OPT_U32, 64<<10)
// ext4 allows 4k xattrs total including some smallish extra fields and the
// keys.  We're allowing 2 512 inline attrs in addition some some filestore
// replay attrs.  After accounting for those, we still need to fit up to
// two attrs of this value.  That means we need this value to be around 1k
// to be safe.  This is hacky, but it's not worth complicating the code
// to work around ext4's total xattr limit.
OPTION(filestore_max_xattr_value_size_other, OPT_U32, 1<<10)
```
So they set it to 64KiB for xfs and btrfs and 1KiB for everything else.  It appears it may be overridden by setting the ```filestore_max_xattr_value_size configuration``` parameter (which I have no idea how to do).

**Comment by kevin-olbrich** [comment]
This is set in the global context of /etc/ceph/ceph.conf

But this does not work around the problem, it's just to disable the warning. Transactions which can not handle this special situation will just fail with unknown results.

**Comment by dweeezil** [comment]
@kevin-olbrich Then it sounds like we'll need an strace to see what call is failing or at least some sort of error from ceph which would help track down the failing syscall.

**Comment by RubenKelevra** [comment]
@kevin-olbrich: any progress?

**Comment by nlienard** [comment]
On my side;  i had to add in ceph.conf for each osd using ZFS : 

osd max object name len = 256
osd max object namespace len = 64

Otherwise; osd won't start as described in this issue.

**Comment by RubenKelevra** [comment]
@nlienard right, this is also our workaround, but I'm not sure about the performance/reliability-impact of this options.

Thanks for your confirmation.

**Comment by behlendorf** [comment]
@nlienard @RubenKelevra this looks like an issue the Ceph developers need be aware of and could resolve.  Is their an open issue in their tracker for this?

**Comment by RubenKelevra** [comment]
@behlendorf actually that's the workaround for filesystems with these limitation. So there was a change in ceph which increases the usage of XAttr and Filenames which exeeds the provided properties of zfs. I don't know exactly what info isn't able to be stored on zfs, but this settings above tells the ceph daemon that it has to be somewhat backwards compatible and work around this limitation.

**Comment by behlendorf** [comment]
My understanding from https://github.com/zfsonlinux/zfs/issues/4913#issuecomment-236889312 was that Ceph was hardcoding the limits based on filesystem type.  Since zfs isn't in their list of known filesystems they just fallback to assuming the worst case.  I'm not familiar with the Ceph code but it looks to me like the missing logic belongs [right here](https://github.com/ceph/ceph/blob/master/src/os/filestore/FileStore.cc#L5643).

**Comment by RubenKelevra** [comment]
@behlendorf so you say on the one hand ceph wants to save large entrys on the filesystem and on the other hand there are no detection of the actual limit and they assume a to small default-limit for the large entrys they try to safe so we need to reduce them at the moment.

So actually we only need to expand the detection to zfs and the error would go away?

**Comment by behlendorf** [comment]
@RubenKelevra yes, that's my quick reading of the Ceph code in question.  On the ZFS side we don't have any meaningful limits of the size or number of xattrs which can be stored.

**Comment by GeLiXin** [comment]
@behlendorf  I have discussed this issue with a core member of Ceph today, he said there is no problem to use ZFS as back-end storage technically, but ZFS is too heavy for them. 

Ceph implement redundancy protection and value-adding functions by themselves, what they want is a simple and high-performance local filesystem. And they are going to implement their own filesystem now.

This seems not a good news for us.

**Comment by johnkeates** [comment]
@GeLiXin it does make sense, since Ceph does the reliability and integrity one level above the FS. Maybe the only way to go beyond that is a cluster-ZFS?

**Comment by kpande** [comment]
if they implement their own filesystem, it can go on ZVOL, right?

**Comment by johnkeates** [comment]
@kpande would XFS-on-ZVOL be any different in that case?

**Comment by jasonrm** [comment]
@GeLiXin @johnkeates my understanding from [a ceps-users thread][0] is that Ceph does basic checksum comparison only during a deep scrub. Also, it leaves the repair to the system administrator, [which doesn't look fun][1]. I haven't been able to determine if there is any realtime protection against corruption. Bluestore, the future backend for Ceph OSDs, doesn't seem to address the corruption concern either.

I'm only dealing with data in the TBs and growing slowly, so I can get by with mirrored vdevs. Redundancy at both the ZFS level and Ceph layer, at the cost of available storage being cut in half, is worth it to me.

[0]: http://lists.ceph.com/pipermail/ceph-users-ceph.com/2016-March/008422.html
[1]: http://www.sebastien-han.fr/blog/2015/04/27/ceph-manually-repair-object/

**Comment by RubenKelevra** [comment]
Since ceph also allows very flexible raid-levels where you select the part number, the number of xor parts which is for raid5 1 and raid6 2 etc. it also allows to add a part which reduce calculation complexity on recovery without adding a redundant part.

This is very nice on larger clusters I guess, but currently ceph has some disadvantages running on ZFS: write performance is very low unless you tweak the ceph setup. Ceph writes all received write-requests to a ringbuffer which is instantly synced. After this write-request to the backend storage, the ceph client get it's ack back. Now the ringbuffer is flushed to the ZFS.

Ceph knows two different operation, parallel and sequencing. My description covers sequencing, but as far as I understood Ceph select parallel on ZFS, which issues a lot of sync writings for one write-request. Else the write operations seems to be split to heavily random 4K writes. So SSD operations and HDD operations are very slow. Only enterprise SSDs from intel are capable of handling this type of workload. 

It is possible to split the workload, so the journal is written to one device and the data storage to a different, but the issue with sync-writes remains... so you need a expensive SSD for the journal. And the backend is flooded with random writes at 4K-level as well.

So the first issue is, there seem to be no split between async writes and sync writes for ceph, so zfs is facing massively writes to the journal, if it's written directly to zfs. I guess there's no reason to write a journal, this is only needed if it's likely that the underlying filesystem is loosing data on outages. But ZFS is atomic, only one state is possible, the old or the modified data is given back after a power outage. So normally it should be safe to assume that the journal is not needed.

The problem on this is, the journal from ceph also keeps track on the order of writes. I don't know what happens if zfs get different async write operations - do zfs maintain the order probably in any event? If not the integrity of ceph might be not given without a journal.

Further you have to have a journal, so you have to create a RAM-Disk for this if you plan to use ceph on ZFS "without" a journal.

It would be great if ceph would get a mode to rely on zfs for it's atomic writes and doesn't create a journal. 
And it would be nice if it would be possible to use async writes thru ceph from the upper block level for VMs down to zfs, as far as observed by myself this seems not the case at the moment - everything is written to the journal.

**Comment by x041** [comment]
@jasonrm there was a lot of debate over checksuming in ceph back in april/may, but ultimately it looks like they decided to implement full checksum on write AND read. In fact, since July bluestore on master branch has full checksum on-read enabled by default. I haven't had a chance to read-over their `csum` implementation, so I don't know if it fully addresses concerns such as the ones we have (silent data corruption) in its current implementation, but its a damn good start from CEPH project.

----

@RubenKelevra There are designs that work quite well with ZFS and CEPH together addressing the journal concern - there is also a (ceph) branch where you enable parallel writes for ZFS (eliminating the ceph journal), but I would actually advice against this, for 2 reasons;
1. working with double journals is actually easier AND higher performing. 
2. the code is not maintained and likely extremely out-of-date

Important note (because I don't want anyone to get the wrong idea) - DO NOT TRY AND BYPASS THE JOURNAL BY USING A RAMDISK - CEPH REQUIRES the journal in WAL (default mode). Even if CEPH guaranteed async writes are properly ordered outside the journal (iirc it does not, but don't quote me on this), it CANNOT automatically recover from journal "disappearing" or being wiped (which would occur with any power cycle with a ramdisk). It is possible to manually recover from "disappearing" journal, but your basically saying "I don't care about integrity of my data", in which case, why even bother with ZFS?

Feel free to contact me directly or via ceph-users (as this is off topic for ZFS issue pages) if you want to discuss this further (CEPH on-top of ZFS design)

---- 

@kevin-olbrich @nlienard @GeLiXin - @jasonrm appears to have already stumbled upon the correct solution and posted it publicly via gist here https://gist.github.com/jasonrm/cec75769098792df056a5a8ee155e6bc . Based on my understanding of CEPH's code base, those parameter changes (specifically `filestore_max_inline_xattr_size = 65536` and `filestore_max_xattr_value_size = 65536`) will change the behavior of CEPH to treat underlying filesystem similar to how it treats XFS, and not just mask it (like setting only `osd max object name len` would) .

Setting xattr=sa is also important from a performance standpoint.

@behlendorf Since this isn't a ZFS issue, I vote to close this issue

**Comment by behlendorf** [comment]
@x041 thanks for posting a clear summary of the situation.  I agree this isn't a ZFS issue and it should be closed, but let's add a section to the [wiki](https://github.com/zfsonlinux/zfs/wiki/FAQ) describing the optimal ZFS filesystem properties for performance and the required CEPH parameters.  Can you verify that the following setting are all that's required.

```
zfs create -o xattr=sa -o dnodesize=auto -o mountpoint=/var/lib/ceph rpool/ceph
```

```
[osd]
filestore_max_inline_xattrs = 10
filestore_max_inline_xattr_size = 65536
filestore_max_xattr_value_size = 65536
```

**Comment by x041** [comment]
@behlendorf Basically, yes. `xattr=sa` and `dnodesize=auto` plus the 3 ceph.conf values are all that's needed to get ZFS working with CEPH in a reasonable fashion.

^^ That's the TLDR. Now I apologize in advance, as I'm about to turn the short response above.... needlessly wordy, heh. Feel free to copy-pasta as desired.

----

There is a lot of tuning that can be done that's dependent on the workload that is being put on CEPH/ZFS, as well as some general guidelines. Some are as follow;

## ZFS

The CEPH filestore back-end heavily relies on xattrs, for optimal performance all CEPH workloads will benefit from the following ZFS dataset parameters
* `xattr=sa` 
* `dnodesize=auto`

Beyond that typically rbd/cephfs focused workloads benefit from small recordsize({16K-128K), while objectstore/s3/rados focused workloads benefit from large recordsize (128K-1M).

## CEPH (ceph.conf}

Additionally CEPH sets various values internally for handling xattrs based on the underlying filesystem. As CEPH only officially supports/detects XFS and BTRFS, for all other filesystems it falls back to rather [limited "safe" values](https://github.com/ceph/ceph/blob/4fe7e2a458a1521839bc390c2e3233dd809ec3ac/src/common/config_opts.h#L1125-L1148). On newer releases need for larger xattrs will prevent OSD's from even starting. 

The officially recommended workaround ([see here](http://docs.ceph.com/docs/jewel/rados/configuration/filesystem-recommendations/#not-recommended)) has some severe downsides, and more specifically is geared toward filesystems with "limited" xattr support such as ext4.

ZFS does not have a limit internally to xattrs length, as such we can treat it similarly to how CEPH treats XFS. We can set overrides to set 3 internal values to the same as those used with XFS([see here](https://github.com/ceph/ceph/blob/9b317f7322848802b3aab9fec3def81dddd4a49b/src/os/filestore/FileStore.cc#L5714-L5737) and [here](https://github.com/ceph/ceph/blob/4fe7e2a458a1521839bc390c2e3233dd809ec3ac/src/common/config_opts.h#L1125-L1148)) and allow it be used without the severe limitations of the "official" workaround. 

```
[osd]
filestore_max_inline_xattrs = 10
filestore_max_inline_xattr_size = 65536
filestore_max_xattr_value_size = 65536
```

## Other General Guidelines

* Use a separate journal device. Do not don't collocate CEPH journal on ZFS dataset if at all possible, this will quickly lead to terrible fragmentation, not to mention terrible performance upfront even before fragmentation (CEPH journal does a dsync for every write).
* Use a SLOG device, even with a separate CEPH journal device. For some workloads, skipping SLOG and setting `logbias=throughput` may be acceptable.
* Use a high-quality SLOG/CEPH journal device, consumer based SSD, or even NVMe WILL NOT DO (Samsung 830, 840, 850, etc) for a variety of reasons. CEPH will kill them quickly, on-top of the performance being quite low in this use. Generally recommended are [Intel DC S3610, S3700, S3710, P3600, P3700], or [Samsung SM853, SM863], or better.
* If using an high quality SSD or NVMe device(as mentioned above), you CAN share SLOG and CEPH Journal to good results on single device. I use a ratio of 4 HDDs to 1 SSD (Intel DC S3710 200GB), with each SSD partitioned (remember to align!) to 4x10GB (for ZIL/SLOG) + 4x20GB (for CEPH journal) - to very good results. 

Again - CEPH + ZFS will KILL a consumer based SSD VERY quickly. I can promise you, even ignoring the lack of power-loss protection, and endurance ratings, you will be very disappointed with performance of consumer based SSD under such a workload.

**Comment by rincebrain** [comment]
One question for you, @x041 - what will that do to an existing configuration with the official workaround enabled?

Because I tried that on a single OSD of an existing configuration with that workaround enabled, and now the OSD won't start with or without those config lines with a nice os/filestore/FileJournal.h: 440: FAILED assert(fd == -1) in the log.

(Putting this in here just so other people who might already be running this don't make whatever mistake I made in presuming this could be done to an existing config as I did.)

**Comment by x041** [comment]
@rincebrain the ceph.conf changes should be safe to implement on existing ZFS/CEPH installations running recent versions. 

What version of CEPH and ZFS are you running, on what OS?

**Comment by maz** [comment]
@x041 The logs are at https://gist.github.com/anonymous/18f243bbb85cdbe159c9c62224c08130
We're running, on Debian jessie:
`ceph version 10.2.2 (45107e21c568dd033c2f0a3107dec8f0b0e58374)`
and ZOL from jessie-backports: `0.6.5.8-1~bpo8+1`

**Comment by x041** [comment]
@maz sorry, just to avoid confusion, are you and @rincebrain referring the same system? Or are you running into a similar issues on a different system?

**Comment by maz** [comment]
@x041 Yes. Sorry. @rincebrain and I are working on the same system 😄

**Comment by x041** [comment]
@maz @rincebrain Looks like on your system that the journal is collocated on ZFS volume, and the block size (recordsize) of the journal file is being detected as 1M (which is ZFS default), meanwhile CEPH is expecting the journal to have a block size (recordsize) of `4096`

```
    -2> 2016-12-20 12:52:53.983907 7f813d19a800  2 journal open journal block size 1048576 != current 4096
-1> 2016-12-20 12:52:53.983953 7f813d19a800 3 journal journal_replay open failed with (22) Invalid argument
```

I'm gonna try and recreate the issue, but in the meanwhile, could you try setting `journal_block_size = 1048576` in `ceph.conf` under `[osd.1]` (set it just for that single OSD, don't set it globally) and let me know how that works.

**Comment by maz** [comment]
@x041: that fixed it :)

**Comment by behlendorf** [comment]
@x041 thanks for the write up!  I've added it almost verbatim to the [wiki](https://github.com/zfsonlinux/zfs/wiki/FAQ#cephzfs) for easy reference.  Closing this issue.

**Comment by RubenKelevra** [comment]
@x041 I would not recommend to use a SSD for Ceph's journal but use a single battery backup per node, which shut off the node in case of a power loss and use a ram-disk for Ceph. Writing all Data to an SSD with DSync and then to a ZFS storage makes no sense at all. 

I would rather use the SSD for ZFS read cache, that's much less wasteful usage of the hardware. :)

**Comment by johnkeates** [comment]
@RubenKelevra wouldn't NV-DIMM be a right fit for that use case?

**Comment by x041** [comment]
@RubenKelevra A ram-disk is an extremely irresponsible recommendation/choice for most (99%) use-cases and end-users. For most users out where an all-flash system does not make sense, offloading journals/logs to a SSD is the most efficient choice between cost, performance, and durability.

For users chasing extreme performance, as @johnkeates mentioned, a NVRAM-NVMe or NVRAM-NVDIMM is a much better and reliable/durable fit.

Basically breaks down to the old quote - "cheap, fast, expensive - pick 2". If you have a design and use-case where makes sense, I hope it works out for you. However it is not a good general recommendation to make.
